clippy_lints/src/methods/mod.rs

```diff
@@ -3378,6 +3379,36 @@ declare_clippy_lint! {
     "calling `Stdin::read_line`, then trying to parse it without first trimming"
 }
 
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for usage of `.map(|_| format!(..)).collect::<String>()`.
+    ///
+    /// ### Why is this bad?
+    /// This allocates a new string for every element in the iterator.
+    /// This can be done more efficiently by creating the `String` once and appending to it using `Iterator::fold`.
+    ///
+    /// ### Example
+    /// ```rust
+    /// fn hex_encode(bytes: &[u8]) -> String {
+    ///     bytes.iter().map(|b| format!("{b:02X}")).collect()
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// use std::fmt::Write;
+    /// fn hex_encode(bytes: &[u8]) -> String {
+    ///     bytes.iter().fold(String::new(), |mut output, b| {
+    ///         let _ = write!(output, "{b:02X}");
+    ///         output
+    ///     })
+    /// }
+    /// ```
+    #[clippy::version = "1.72.0"]
+    pub FORMAT_COLLECT,
+    perf,
```

Comment:
`write!` never errors on `String`s, so I'm not too sure on that. But yes, I agree that this might be too annoying for `perf`.

Reply:
Interesting, that should probably be documented in the lint description with a comment then as that's rather unexpected behavior (at least from the reader's perspective)

Reply:
I guess it's only really a useful suggestion if the iterator is gonna yield lots of small items, like bytes or chars (which seems hard to put into a lint, other than maybe restricting it to small types), otherwise it feels pedantic, like here: https://github.com/rust-lang/rust-clippy/pull/11116/files#diff-42cb6807ad74b3e201c5a7ca98b911c5fa08380e942be6e4ac5807f8377f87fcL135

Reply:
> `write!` never errors on `String`s, so I'm not too sure on that. But yes, I agree that this might be too annoying for `perf`.

`Display` impls can return errors even if `String` does not (which is not advisable though)